src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,283 @@
+                  partition.replicas->begin(),
+                  partition.replicas->end(),
+                  [](const model::node_id& node_id) { return node_id < 0; });
+                if (negative_node_id_it != partition.replicas->end()) {
+                    vlog(
+                      klog.error,
+                      "Negative broker id in replica assignment: topic {}, pid "
+                      "{}, replicas {}",
+                      topic.name,
+                      partition.partition_index,
+                      *partition.replicas);
+                    resp.data.error_code
+                      = error_code::invalid_replica_assignment;
+                    resp.data.error_message = fmt::format(
+                      "Invalid broker id in replica list: {}",
+                      *partition.replicas);
+                    co_return co_await ctx.respond(std::move(resp));
+                }
+
+                auto all_node_ids = ctx.metadata_cache().node_ids();
+                auto unkown_broker_id_it = std::find_if(
+                  partition.replicas->begin(),
+                  partition.replicas->end(),
+                  [all_node_ids](const model::node_id& node_id) {
+                      return std::find(
+                               all_node_ids.begin(),
+                               all_node_ids.end(),
+                               node_id)
+                             == all_node_ids.end();
+                  });
+                if (unkown_broker_id_it != partition.replicas->end()) {
+                    vlog(
+                      klog.error,
+                      "Broker Id is not in the cluster: topic {}, pid {}, "
+                      "broker id "
+                      "{}, live brokers: {}",
+                      topic.name,
+                      partition.partition_index,
+                      *unkown_broker_id_it,
+                      all_node_ids);
+                    resp.data.error_code
+                      = error_code::invalid_replica_assignment;
+                    resp.data.error_message = fmt::format(
+                      "Replica assignment has brokers that are not alive. "
+                      "Replica list: {}, live broker list: {}",
+                      *partition.replicas,
+                      all_node_ids);
+                    co_return co_await ctx.respond(std::move(resp));
+                }
+
+                auto assignment_exists
+                  = co_await ctx.topics_frontend().replica_set_exists(
+                    ntp, *partition.replicas);
+                if (assignment_exists) {
+                    // The requested reassignment already exists, so skip this
+                    // one
+                    vlog(
+                      klog.info,
+                      "Request to change ntp {} replica set to {}, no change",
+                      ntp,
+                      *partition.replicas);
+                    continue;
```

Comment:
Shouldn't the response object be filled with some error code like `partition_assignment_already_exists`? Theres one for topics.

Reply:
I could not find such an error in the [Kafka protocol](https://kafka.apache.org/protocol.html#protocol_error_codes). How about returning a retriable error code such as `unknown_topic_or_partition` with a clear message.

FYI: This check isn't performed in Apache Kafka. We do this check on the admin_server to avoid unnecessary work on the Redpanda side. So we could omit this check. Thoughts?

```cpp
    auto current_assignment
      = _controller->get_topics_state().local().get_partition_assignment(ntp);

    // For a no-op change, just return success here, to avoid doing
    // all the raft writes and consensus restarts for a config change
    // that will do nothing.
    if (current_assignment && current_assignment->replicas == replicas) {
        vlog(
          logger.info,
          "Request to change ntp {} replica set to {}, no change",
          ntp,
          replicas);
        co_return ss::json::json_void();
    }
```

Reply:
Unless we have a strong reason not to, we should behave the same as Kafka. If we have a case that doesn't exist in Kafka, then let's state that explicitly and discuss. The source of truth for behavior is not the KIP, it is the Kakfa implementation. If it would be helpful to have a quick overview of how to navigate that, I can help just ping me on slack.

Reply:
The admin server is a place we can kind of have our own freedom because we don't have to be compliant to another protocol, but when responding to kafka clients we should behave and respond to client as kafka would. I agree with Noah, lets just see what kafka does and copy it.